Use REST API for payment requests

luukveenis opened the pull request:
I ran into an issue recently where I needed to provide data along with payment requests that was not supported by the underlying Adyen gem. It turned out to be very difficult to customize the request parameters using the SOAP API implementation we currently rely on. The implementation consists largely of chunks of XML as strings appended together with data injected into them using string formatting. [Example here](https://github.com/wvanbergen/adyen/blob/master/lib/adyen/api/templates/payment_service.rb)

The Adyen gem also provides an implementation of the REST API that posts the provided params as form data. This change replaces all our API calls to use that implementation instead. Since we can provide arbitrary parameters in the request, this should make it easier to customize the request for different payment methods.

luukveenis committed:
Use REST API for initial authorization request

We want to slowly move everything onto the REST API. The way the SOAP
implementation works in the Adyen gem is scary and nearly impossible to
customize or work with.

Commit:
Migrate remaining payment requests to REST API

For the various calls there are options that we may want to provide that
aren't easily supported by the Adyen gem. One example is for the
openinvoice payment method, we need to provide invoice line data along
with partial captures and refunds.

Since the SOAP API implementation is just a collection of XML strings
concatenated together, it's very hard to customize this to provide
additional data. Using the REST API, we just provide a hash of
attributes with the payment request where we can stick whatever data we
need to.

Commit:
Fix failing specs due to new api

Most of the spec failures were caused by the mock API. This updates the
mock API to return responses representative of those from the REST API.

Commit:
Improve error handling for API requests

Some of the calls that previously returned 200 responses with an error
message using the SOAP API, now cause 5xx errors through the REST API.
This catches those errors and throws a gateway error with the
corresponding message so we can process it in Solidus.

Commit:
Allow additional data to be passes to #credit

The additional data container is used by Adyen to provide any extra
information required for various payment methods.

One example use case is for partial refunds on invoice payments they
require us to provide information about the invoice lines being modified
in this container.

Commit:
Pass full gateway options to credit!

All other gateway actions receive the full gateway options, there's no
reason we shouldn't provide it here. The currency is included in the
gateway options by default.

This is also done so that we can use the gateway options to provide
additional data for Adyen. One example is to provide invoice line
information when performing partial refunds for invoice type payments.

Dkendal commented on app/models/concerns/spree/gateway/adyen_gateway.rb:
```diff
@@ -21,61 +21,81 @@ def merchant_account
       currency ||= options[:originator].payment.currency
-      amount = { currency: currency, value: amount }
+      params = payment_params(amount, currency, psp_reference)
+      params.merge!(options.slice(:additional_data)) if options[:additional_data]
```
are you sure you want to mutate `params` here?

Dkendal commented on app/models/concerns/spree/gateway/adyen_gateway.rb:
```diff
@@ -21,61 +21,81 @@ def merchant_account
 
     private
 
+    def rest_client
```
This method, and the calls like

``` ruby
+        ::Adyen::REST.session(rest_client) do |client|
 +          client.list_recurring_details({
 +            merchant_account: payment_method.merchant_account,
 +            shopper_reference: reference_number_from_order,
 +          })
```

have a bit of duplication.

What do you think of moving them to a separate class to provide an abstraction over the rest client?

Commit:
Extract API interactions into their own class

This logic was repeated in multiple places and made more sense to keep
in one class.

Dkendal commented on app/models/concerns/spree/adyen/payment.rb:
```diff
@@ -125,30 +129,20 @@ def process &block
-            )
-          end
+        # If this is a new credit card we should have the encrypted data
+        if source.encrypted_data
```
would be nice if we had a predicate defined on the source for this. The fact that the source has encrypted data does not imply to me, as an outsider, that this is a new credit card.

Something like `source.new_credit_card?` would be nice.

Dkendal commented on app/models/concerns/spree/adyen/payment.rb:
```diff
@@ -125,30 +129,20 @@ def process &block
+        # If this is a new credit card we should have the encrypted data
+        if source.encrypted_data
+          rest_client.authorise_recurring_payment(payment_params.merge(encrypted_card_data))
+        elsif source.gateway_customer_profile_id
```
likewise here, that fact that the source has a gateway customer profile id doesn't mean anything to me and doesn't imply that there should be recurring payment - a predicate on the source that explained the relation would be useful. Maybe `source.has_profile?` or whatever makes sense.

Dkendal commented on app/models/concerns/spree/adyen/payment.rb:
```diff
@@ -125,30 +129,20 @@ def process &block
+        elsif source.gateway_customer_profile_id
+          rest_client.reauthorise_recurring_payment(payment_params)
+        else
+          raise Spree::Core::GatewayError.new(
```
This should be defined as a custom error that inherits from `Spree::Core::GatewayError`.

Dkendal commented on app/models/concerns/spree/adyen/payment.rb:
```diff
@@ -125,30 +129,20 @@ def process &block
-      rescue ::Adyen::REST::ResponseError => error
-        log_entries.create!(details: error.to_yaml)
-        raise Spree::Core::GatewayError.new(error.message)
+      rescue Spree::Core::GatewayError => gateway_error
```
Are you intending to capture all gateway errors here, or only the one created on like 139?

Commit:
Move credit card auth logic onto gateway

Instead of polluting Spree::Payment with more stuff, keep this seperate
on the gateway. I'm passing the payment itself through because we need
access to the order, the source, and various data from the payment.

Commit:
Use predicate methods to improve readability

Commit:
Name helper method more accurately

This method is only used to construct the parameters for modification
requests (captures or refunds). A refund isn't really a payment request,
so this is a bad name for it.

Commit:
Remove unused error and raise custom error

The ClearTextCardNumberError was not used anymore and can be removed.
For the other we want to raise something that better describes what's
going on. By inheriting from Spree::Core::Gateway error, Solidus will
catch it and handle it appropriately.

Commit:
Create wrapper classes for REST API responses

The REST API implementation in the Adyen gem can return a number of
different responses with different public APIs. This makes them
difficult to work with, so this class is intended to make it easier to
process the responses.

Commit:
Rename mock adyen api to mock adyen client

I changed this context to stub out the Spree::Adyen::Client class
instead, which makes all the API calls. The name wasn't an accurate
description of the context anymore, so this updates it to reflect that
change.

Dkendal commented on app/models/spree/gateway/adyen_credit_card.rb:
```diff
@@ -19,5 +20,125 @@ def method_type
+
+      unless response.success?
+        payment.log_entries.create!(details: response.to_yaml)
+        raise Spree::Core::GatewayError.new(
```
Should raise a custom error instead. something that inherits from Gateway error, that way you don't have to explicitly set the error message as well.

Dkendal commented on app/models/spree/gateway/adyen_credit_card.rb:
```diff
@@ -19,5 +20,125 @@ def method_type
+        )
+      else
+        raise EncryptedDataError.new(
+          I18n.t(:missing_encrypted_data, scope: 'solidus-adyen')
```
initializer for the error class should set this message, not the caller.

Dkendal replied:
also I don't think it appropriate for errors to use i18n, errors are not intended to be user facing and there for should not use localizations.

Dkendal commented on app/models/spree/gateway/adyen_credit_card.rb:
```diff
@@ -19,5 +20,125 @@ def method_type
+        response.gateway_response.details
+      else
+        raise ProfileLookupError.new(
+          I18n.t(:profile_lookup_failed, scope: "solidus-adyen")
```
see previous comment.

Dkendal commented on lib/spree/adyen/api_response.rb:
```diff
@@ -0,0 +1,61 @@
+      end
+
+      def psp_reference
+        @gateway_response[:psp_reference]
```
should use `fetch!` instead.

Dkendal commented on lib/spree/adyen/api_response.rb:
```diff
@@ -0,0 +1,61 @@
+
+      def error_message
+        if authorisation_response?
+          @gateway_response[:refusal_reason]
```
see previous.

Dkendal commented on lib/spree/adyen/api_response.rb:
```diff
@@ -0,0 +1,61 @@
+        elsif error_response?
+          @gateway_response.message
+        else
+          I18n.t(:unknown_gateway_error, scope: "solidus-adyen")
```
see earlier comment.

Commit:
Provide default error messages for custom errors

Since we always want the message to be the same, they should be set on
the class itself rather than being passed in. Overriding #message rather
than the constructor allows us to raise the error without any arguments
and still set the custom message.

Commit:
Update change log
